### PR TITLE
Use system properties and declared some suggested static properties final

### PR DESCRIPTION
**Pull request by StefanSpieker**
Use system properties and declared some suggested static properties final

Follow up of #5235 
### Proposed changelog entries

* Internal: Declared some internal fields to final. Should not impact any existing script console commands.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by basil — COMMENTED** [review]
I haven't reviewed this change closely, but I wanted to draw attention to the fact that making `public static` constants `final` means that the compiler may begin to [inline](https://stackoverflow.com/questions/3524150/is-it-possible-to-disable-javacs-inlining-of-static-final-variables) these constants when compiling _consumers_ (see "[The Static Final Inline Trap](https://javasplitter.blogspot.com/2011/10/static-final-inline-trap.html)" and "Puzzle 93: Class Warfare" in Joshua Bloch's _Java Puzzlers_). This has some implications on binary compatibility, especially if we want to change these values in the future. For example, if a plugin compiled against a core with these changes inlines a `public static final` constant when it is compiled, then changing the value in a future core release would not be enough for the plugin to begin using the new value; the plugin would rather need to be recompiled. Depending on the use case, this could be a problem.

This may or may not be a problem in practice, but I wanted to draw attention to this phenomenon to ensure that it is considered as part of the due diligence for this change. If it is considered and deemed not to be a problem in practice, then that would be fine by me.

**Review by basil — COMMENTED** [review]
Creating `/tmp/additionalFields` with

```
hudson.scm.SCMRevisionState#NONE
hudson.security.GlobalSecurityConfiguration#FILTER
hudson.util.ChartUtil#CHEBYSHEV_N
hudson.util.ColorPalette#LINE_GRAPH
hudson.util.io.ArchiverFactory#TAR
hudson.util.io.ArchiverFactory#TARGZ
hudson.util.io.ArchiverFactory#ZIP
jenkins.install.SetupWizard#initialSetupAdminUserName
jenkins.model.IdStrategy#CASE_INSENSITIVE
jenkins.tools.GlobalToolConfiguration#FILTER
jenkins.util.xstream.XStreamDOM#REPLACER
```

then using jenkins-infra/usage-in-plugins to look for usages in plugins, including those in CloudBees CI, with

```
mvn process-classes exec:exec -Dexec.executable=java -Dexec.args='-classpath %classpath org.jenkinsci.deprecatedusage.Main --additionalFields /tmp/additionalFields --onlyIncludeSpecified --updateCenter https://jenkins-updates.cloudbees.com/update-center/envelope-core-oc/update-center.json?version=2.277.2.3,https://jenkins-updates.cloudbees.com/update-center/envelope-core-mm/update-center.json?version=2.277.2.3'
```

produced the following references:

#### `hudson/scm/SCMRevisionState#NONE`

- accurev
- bazaar
- cmvc
- darcs
- dimensionsscm
- filesystem_scm
- git
- gradle-repo
- harvest
- integrity-plugin
- multiple-scms
- plasticscm-mergebot
- plasticscm-plugin
- repo
- teamconcert
- template-project
- vss

#### `hudson/util/io/ArchiverFactory#TAR`

- operations-center-context

#### `hudson/util/io/ArchiverFactory#TARGZ`

- artifact-manager-s3
- azure-artifact-manager
- google-cloudbuild
- jdcloud-codedeploy
- shelve-project-plugin
- workflow-api

#### `hudson/util/io/ArchiverFactory#ZIP`

- codebeamer-xunit-uploader
- infradna-backup
- pipeline-utility-steps

#### `jenkins/install/SetupWizard#initialSetupAdminUserName`

- master-provisioning-core

#### `jenkins/model/IdStrategy#CASE_INSENSITIVE`

- ldap
- mock-security-realm
- nectar-rbac

The safest option when it comes to binary compatibility is not to change any `public static` fields to `public static final` fields if there are existing references from plugins. This way we guarantee that plugins are always using the latest value of the field from Jenkins core rather than inlining a possibly older value. So I would recommend removing the `final` keyword from the fields that turned up in the search above:

- `hudson/scm/SCMRevisionState#NONE`
- `hudson/util/io/ArchiverFactory#TAR`
- `hudson/util/io/ArchiverFactory#TARGZ`
- `hudson/util/io/ArchiverFactory#ZIP`
- `jenkins/install/SetupWizard#initialSetupAdminUserName`
- `jenkins/model/IdStrategy#CASE_INSENSITIVE`

The following fields are safe to add the `final` keyword to since they are not _currently_ used by any plugins:

- `hudson.security.GlobalSecurityConfiguration#FILTER`
- `hudson.util.ChartUtil#CHEBYSHEV_N`
- `hudson.util.ColorPalette#LINE_GRAPH`
- `jenkins.tools.GlobalToolConfiguration#FILTER`
- `jenkins.util.xstream.XStreamDOM#REPLACER`

But we should also take care to prevent any plugins from starting to use these fields in the future by adding `@Restricted(NoExternalUse.class)` and `@RestrictedSince("2.TODO")`. I would also hesitate to restrict `hudson.util.ColorPalette#LINE_GRAPH` since it seems actually useful to plugins. But `hudson.security.GlobalSecurityConfiguration#FILTER`, `hudson.util.ChartUtil#CHEBYSHEV_N`, `jenkins.tools.GlobalToolConfiguration#FILTER`, and `jenkins.util.xstream.XStreamDOM#REPLACER` seem only useful from within Jenkins core, so I think it makes sense to restrict them and then mark them final.

**Comment by StefanSpieker** [comment]
Thanks @basil for the thorough review. 👍 I changed it as suggested.

File: core/src/main/java/hudson/util/ColorPalette.java
Comment (basil) [inline core/src/main/java/hudson/util/ColorPalette.java:50]: Technically not "used in several plugins" because no plugins are _currently_ using this field, but I think it makes sense to keep it non-final because it seems useful to expose to plugins. So I would update the justification to read: "exposed to plugins".

File: core/src/main/java/jenkins/util/xstream/XStreamDOM.java
Comment (basil) [inline core/src/main/java/jenkins/util/xstream/XStreamDOM.java:548]: I guess I forgot to search for this one with `usage-in-plugins`. Can you verify this is unused with `usage-in-plugins`? If that is asking too much, I can look again when I have some time.

**Review by basil — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
